### PR TITLE
[WIP] Ft new ineq structure

### PR DESCRIPTION
**Pull request by LuziaKn**
Starting a new setup for defining inequality constraints in a more modular way.

The goal is to easily allow the selection of certain constraint types and improve the readability of files.
Furthermore, the newest urdfenvs version is now used.

### CODE REVIEW

**Comment by maxspahn** [comment]
We must remove all the urdf related files from this repo, that means the meshes and images. It should be sufficient to use the urdfenvs

**Comment by maxspahn** [comment]
Which version of urdfenvs does it depend on? path `vendor/gym_envs_urdf` cannot be found.

File: examples/makeSolver.py [hunk 1]
```diff
@@ -12,6 +12,7 @@ def parse_setup(setup_file: str):
 
 def main(robot_type, setup_file):
     setup = parse_setup(setup_file)
+    setup['robot']['urdf_file'] = current_path = os.path.dirname(os.path.abspath(__file__)) + "/assets/"+str(robot_type)+"/" + setup['robot']['urdf_file']
```
Comment (maxspahn) [inline examples/makeSolver.py:15]: what is the variable `current_path` being used for?

File: examples/makeSolver.py [hunk 2]
```diff
@@ -22,8 +23,13 @@ def main(robot_type, setup_file):
...
+    """ Commented out by Saray:
+    if len(sys.argv) < 2 and not input:
         print("Please provide a config file for solver generation.")
+        sys.argv.append('config/' + str(robot_type) + "Mpc.yaml")
     else:
-        robot_type = re.findall('\/(\S*)M', sys.argv[1])[0]
-        main(robot_type, sys.argv[1])
+    robot_type = re.findall('\/(\S*)M', sys.argv[1])[0]
+    """
+    setup_file = 'config/' + str(robot_type) + "Mpc.yaml"
```
Comment (maxspahn) [inline examples/makeSolver.py:34]: Why did you remove this functionality of passing the config file through the command line?

File: examples/panda_example.py [hunk 3]
```diff
@@ -1,33 +1,64 @@
...
+import os
 import numpy as np
-import gym
-import urdfenvs.panda_reacher
-from MotionPlanningEnv.sphereObstacle import SphereObstacle
-from MotionPlanningGoal.staticSubGoal import StaticSubGoal
+import gymnasium as gym
+# import urdfenvs.panda_reacher
+# from MotionPlanningEnv.sphereObstacle import SphereObstacle
+# from MotionPlanningGoal.staticSubGoal import StaticSubGoal
```
Comment (maxspahn) [inline examples/panda_example.py:7]: These lines can be removed instead of commented.

File: examples/panda_example.py
Comment (maxspahn) [inline examples/panda_example.py:1]: Panda example is now broken due to the velocity weights required as list instead of scalar value.

File: examples/pointRobot_example.py
Comment (maxspahn) [inline examples/pointRobot_example.py:1]: This example is also broken because of required list for velocity weights.

File: examples/pointRobot_example_old.py
Comment (maxspahn) [inline examples/pointRobot_example_old.py:1]: It is very bad pracice to have dead files like old versions in the repositiory. 
Make sure that one of the two pointRobot examples work and remove the other one.

File: robotmpcs/models/mpcModel.py [hunk 4]
```diff
@@ -1,87 +1,54 @@
+import os
 import casadi as ca
 from dataclasses import dataclass
 from forwardkinematics.fksCommon.fk import ForwardKinematics
 from forwardkinematics.urdfFks.generic_urdf_fk import GenericURDFFk
 import numpy as np
+import sys
+sys.path.append("../")
+sys.path.append("")
```
Comment (maxspahn) [inline robotmpcs/models/mpcModel.py:9]: Are these really neceassary?

File: robotmpcs/models/mpcModel.py [hunk 5]
```diff
@@ -1,87 +1,54 @@
...
+        self._kwargs = kwargs
+
         if initParamMap:
             self._limits = {
                 "x": {"low": np.ones(self._nx) * -100, "high": np.ones(self._nx) * 100},
                 "u": {"low": np.ones(self._nu) * -100, "high": np.ones(self._nu) * 100},
                 "s": {"low": np.zeros(1), "high": np.ones(1) * np.inf},
             }
             self.initParamMap()
+        inequality_list = ['obstacle_avoidance', 'speed_limits', 'joint_limits', 'input_limits', 'self_collision']
```
Comment (maxspahn) [inline robotmpcs/models/mpcModel.py:41]: `speed` and `vel` are now used interchangable. We should be consistent on that.

File: robotmpcs/planner/mpcPlanner.py [hunk 6]
```diff
@@ -2,13 +2,22 @@
 import yaml
 import os
 import sys
+sys.path.append("../")
+sys.path.append("")
```
Comment (maxspahn) [inline robotmpcs/planner/mpcPlanner.py:6]: This arbitrary adding of paths is bad practice. the function`load_forces_path` can simply be inside robotmpcs.

File: robotmpcs/planner/mpcPlanner.py [hunk 7]
```diff
@@ -184,10 +218,18 @@ def shiftHorizon(self, output, ob):
             if stage == 1:
                 continue
             self._x0[stage - 2, 0 : len(output[key])] = output[key]
+            self._x0[-1:, 0 : len(output[key])] = output[key]
+
+    def setX0(self, initialize_type="current_state", initial_step= True):
```
Comment (maxspahn) [inline robotmpcs/planner/mpcPlanner.py:223]: Well done.

File: robotmpcs/planner/mpcPlanner.py [hunk 8]
```diff
@@ -222,34 +269,44 @@ def solve(self, ob):
...
+        if self._config.control_mode == "vel":
+            action = self.output[key1][-self._nu-self._nu: -self._nu]
+        elif self._config.control_mode == "acc":
+            action = self.output[key0][-self._nu:]
+        else:
+            print("No valid control mode specified!")
+            action = np.zeros((self._nu))
         if self._config.slack:
-            self._slack = output[key1][self._nx]
+            self._slack = self.output[key1][self._nx]
```
Comment (maxspahn) [inline robotmpcs/planner/mpcPlanner.py:293]: Why id you shift the slack by one step?

**Review by maxspahn — CHANGES_REQUESTED** [review]
THanks @LuziaKn for this effort. There was a lot of work in making robotmpcs compatible with newere versions of urdfenvs. 

Concerning the new structure with the inequality constraints. I think it is very good to have it split up. As you also made changes in all the other parts, we will have to work a bit more on the other parts, such that the panda example and the point_robot are working again.

I also left some comments in the PR.

I can also work on it, so let me know if you are done with your changes, because I would like to do some cosmetics to the PR so that it continues being generic.

File: examples/makeSolver.py (hunk 1 above)
Comment (LuziaKn) [reply examples/makeSolver.py:15]: the variable `current_path` is not used and can be removed.

File: examples/makeSolver.py (hunk 2 above)
Comment (LuziaKn) [reply examples/makeSolver.py]: I changed it back to its original version.

File: examples/panda_example.py (hunk 3 above)
Comment (LuziaKn) [reply examples/panda_example.py]: Done

File: examples/panda_example.py
Comment (LuziaKn) [reply examples/panda_example.py:1]: Panda example is runable now. However, the mpc is currently not feasible. exitflag -7.

File: examples/pointRobot_example_old.py
Comment (LuziaKn) [reply examples/pointRobot_example_old.py:1]: removed

File: robotmpcs/models/mpcModel.py (hunk 4 above)
Comment (LuziaKn) [reply robotmpcs/models/mpcModel.py]: removed

File: robotmpcs/models/mpcModel.py (hunk 5 above)
Comment (LuziaKn) [reply robotmpcs/models/mpcModel.py:31]: This line is not required anymore but I changed `speed` to `vel` to be consistent.

File: robotmpcs/planner/mpcPlanner.py (hunk 6 above)
Comment (LuziaKn) [reply robotmpcs/planner/mpcPlanner.py:6]: removed

File: examples/panda_example.py
Comment (maxspahn) [reply examples/panda_example.py:1]: Fixed by cd5b288

**Comment by maxspahn** [comment]
@LuziaKn Can you have a last look over it?
I have decided to do some cosmetics later, because we are kind of pressured.
Let's merge that quickly and work on the plane constraints and the ros-wrapper tomorrow.

**Comment by maxspahn** [comment]
> Currently, the function set_mpc_parameters is called and defined in the example files. The constraints, e.g. the vellimits , should/can only be added when the paramMap of the solver includes the corresponding variable is available.

The idea is that you use the same config file for the solver generation as for running the example. Then, this should not be a problem, right?
